src/sknnr/transformers/_constrained_transformer.py

```diff
@@ -16,11 +25,14 @@ def fit(self, X, y):
             ensure_min_features=2,
             ensure_min_samples=1,
         )
-        y = np.asarray(y)
-        if len(y.shape) < 2:
-            raise ValueError("`y` must be a 2D array.")
 
-        self.ordination_ = CCA(X, y)
+        if self.method == "cca":
+            self.ordination_ = CCA(X, y)
+        elif self.method == "rda":
+            self.ordination_ = RDA(X, y)
+        else:
+            raise ValueError("`method` must be one of 'cca' or 'rda'.")
```

Comment:
I can see why you put this check here since this is where we need to instantiate the ordination, but if possible it would be nice to make this check in `__init__` so that it fails as soon as possible.

Just to throw out a few untested options, what do you think about something like below?

```python
CONSTRAINED_METHODS = {
    "cca": CCA,
    "rda": RDA,
}

class ConstrainedTransformer(ComponentReducerMixin, TransformerMixin, BaseEstimator):
    def __init__(
        self, method: Literal["cca", "rda"] = "cca", n_components: int | None = None
    ):
        self.method = method
        if self.method not in CONSTRAINED_METHODS:
            raise ValueError(f"`method` must be one of {CONSTRAINED_METHODS.keys()}, not {method}.")
        
        super().__init__(n_components=n_components)

    def fit(self, X, y):
        X = self._validate_data(
            X,
            reset=True,
            dtype=FLOAT_DTYPES,
            force_all_finite=True,
            ensure_min_features=2,
            ensure_min_samples=1,
        )
        self.ordination_ = CONSTRAINED_METHODS[self.method](X, y)
        self.set_n_components()
        self.env_center_ = self.ordination_.env_center
        self.projector_ = self.ordination_.projector(n_components=self.n_components_)
        return self
```

Or we could set the ordination class in `__init__` instead of the name. This is a little cleaner, but maybe it's too weird to take a parameter and set a different attribute with the same name?

```python
class ConstrainedTransformer(ComponentReducerMixin, TransformerMixin, BaseEstimator):
    def __init__(
        self, method: Literal["cca", "rda"] = "cca", n_components: int | None = None
    ):
        self.method = CONSTRAINED_METHODS.get(method)
        if self.method is None:
            raise ValueError(f"`method` must be one of {CONSTRAINED_METHODS.keys()}, not {method}.")
        
        super().__init__(n_components=n_components)

    def fit(self, X, y):
        X = self._validate_data(
            X,
            reset=True,
            dtype=FLOAT_DTYPES,
            force_all_finite=True,
            ensure_min_features=2,
            ensure_min_samples=1,
        )
        self.ordination_ = self.method(X, y)
        self.set_n_components()
        self.env_center_ = self.ordination_.env_center
        self.projector_ = self.ordination_.projector(n_components=self.n_components_)
        return self
```

Reply:
I like either of these patterns ... much better to use a dictionary here.  A couple of questions:

* For constants (dicts or otherwise), I'm never clear when to make them module- or class-level.  Is there an advantage to having `CONSTRAINED_METHODS` as a module-level constant?
* In your second solution, is there a reason why the user argument (`method`) and the attrribute need to be named the same?  Would it make sense to call the attribute something like `self.method_cls`?

I yield to your judgement on what you think is best.  Both solutions are better than I what I currently have and neither is very confusing to me.

Reply:
> Is there an advantage to having CONSTRAINED_METHODS as a module-level constant?

No, I think you're right that this should be class-level!

> In your second solution, is there a reason why the user argument (method) and the attrribute need to be named the same? Would it make sense to call the attribute something like self.method_cls?

I had a vague memory that `sklearn` prevents setting attributes that don't match the `__init__` args, but while [reading up on that](https://scikit-learn.org/stable/developers/develop.html#instantiation), I found that they recommend against having any *logic* in `__init__`:

> There should be no logic, not even input validation, and the parameters should not be changed [in `__init__`]. The corresponding logic should be put where the parameters are used, typically in fit

Looks like you had it right from the beginning!

Reply:
OK, I think I have this right, although it feels weird that we can initialize `ConstrainedTransformer` with a bogus `constrained_method` and it won't error until fitting.  Here's the relevant bits:
```
class ConstrainedTransformer(ComponentReducerMixin, TransformerMixin, BaseEstimator):
    CONSTRAINED_METHODS = {
        "cca": CCA,
        "rda": RDA,
    }

    def __init__(
        self,
        constrained_method: Literal["cca", "rda"] = "cca",
        n_components: int | None = None,
    ):
        self.constrained_method = constrained_method
        super().__init__(n_components=n_components)

    def fit(self, X, y):
        ...
        method_cls = self.CONSTRAINED_METHODS.get(self.constrained_method)
        if method_cls is None:
            raise ValueError(
                f"`method` must be one of {self.CONSTRAINED_METHODS.keys()}, not"
                f" {self.constrained_method}."
            )
        self.ordination_ = method_cls(X, y)
```
Again, if this isn't quite right, I'm happy to change!

Reply:
> it feels weird that we can initialize ConstrainedTransformer with a bogus constrained_method and it won't error until fitting

I agree, but you can do the same thing with some `sklearn` estimators like `criterion` for `RandomForestRegressor`. I guess the downside of building off of `sklearn` is that we inherit some of their questionable choices.

I'm going to suggest one minor change in an inline comment, but otherwise this looks great!